Document ways to get changed courses and providers

benilovj opened the pull request.

benilovj committed:
Document 'Get all changed courses'

benilovj committed:
Document 'Get all changed providers'

misaka commented on source/index.html.md.erb:
```diff
@@ -184,6 +184,34 @@ Parameter | Description
 
+### Get all changed courses
+
+This endpoint retrieves all courses changed since midnight UTC of the specified day.
```
~I don't see how we specify a day. Do you mean the current day the request is made?~

I see it now.

tvararu commented on source/index.html.md.erb:
```diff
@@ -184,6 +184,34 @@ Parameter | Description
+Parameter | Description
+--------- | -----------
+recruitment_cycle | 4-character year (e.g. 2019 for 2019/20 courses)
+changed_since | Day in YYYYMMDD format
```
ISO8601 is `2019-01-11` even though `20190111` is also allowed ([says Wikipedia?](https://en.wikipedia.org/wiki/ISO_8601#Calendar_dates)), [should we support that instead?](https://xkcd.com/1179/) I think it's slightly more readable.

tvararu reviewed (approved):
Can't spot anything wrong with this

misaka commented on source/index.html.md.erb:
```diff
@@ -184,6 +184,34 @@ Parameter | Description
+Parameter | Description
+--------- | -----------
+recruitment_cycle | 4-character year (e.g. 2019 for 2019/20 courses)
+changed_since | Day in YYYYMMDD format
```
Is there a reason we want to provide an interface to get changes with days-granularity? Do we not need/want to support higher-granularity updates? Every hour or N hours, for example?

I like ZenDesk's implementation for chat messages, and we're going to have to do something like that for paging anyway, so why not just do epoch seconds.

tvararu replied:
My 2 cents is the thinking is:

- We're not going to have lots of updates happening, a changeset for a given slow day could be only a few dozen records potentially
- Splitting by days makes it easier to work with for mental arithmetic purposes (thinking in seconds isn't super easy)
- You can cache per day at the database level (though we probably won't need to)

Jake wrote down some more context from a chat we had today though so he might be able to chime in with better reasons.

misaka replied:
@tvararu discussed this some more with @benilovj and we've flipped back to the seconds-resolution thing. I appreciate that this is a lot of flipping back and forth, let's discuss this together. I've updated the documentation with my latest proposal.

benilovj committed:
Increase changed_since granularity to seconds

misaka commented on source/index.html.md.erb:
```diff
@@ -192,6 +192,34 @@ Parameter | Description
 
+### Get all changed courses
+
+This endpoint retrieves all courses changed since midnight UTC of the specified day.
```
Looks like this needs to be updated to reflect the switch to seconds. I can update it.

benilovj replied:
Fixed

misaka committed:
changed_since is now a date/time point in time

Use a changed_since that has resolution in seconds instead of days so that, for
example, on rollover day client(s) don't request the entire years worth of
records multiple times a day.

This also mean implementing a paging system similar to the one used by ZenDesk
for their incremental loading of chat messages:

https://developer.zendesk.com/rest_api/docs/chat/incremental_export

misaka commented:
Having updated the docs just now, I can't help but wonder if there's a reason to include the current paging mechanism (page=1, etc) at all, or whether it shouldn't be completely replaced by the mechanism described in the `Retrieving Changed Records` section I've just added. Would like to have a final discussion about all this if/when we have time today (Tuesday).

benilovj reviewed (commented):
I'm happy with your proposal about the incremental load, @misaka 👍 

Regarding the all records pagination, shall we look at that on another PR to keep things moving?

misaka committed:
Update docs for retrieving changed records

tvararu commented on source/index.html.md.erb:
```diff
@@ -33,24 +33,123 @@ curl "api_endpoint_here"
-Per-Page | The number of results in the page | `Per-Page: 100`
-Total | The total number of results | `Total: 16888`
+The response headers will include information related to continuing to retrieve
+changed records either immediately, or when ready to injest more records.
```
```suggestion
changed records either immediately, or when ready to ingest more records.
```

tvararu committed:
Update source/index.html.md.erb

Fix typo

Co-Authored-By: benilovj <benilov@gmail.com>

tvararu reviewed (approved):
Looks good. 👍

misaka commented:
While making these recent changes, I started wondering if slate really is the right thing to use here. Part of it's core feature-set seems to be to accomodate programming-language APIs where we'd provide usage examples in different languages, which we don't do. To do this it constrains us a little by shoving pre-formatted code blocks to the right of the text instead of allowing them to be inline. We could also get away with a plain MarkDown file that's available via our repository.

Anyone else have similar thoughts?
